%!TEX root = ../thesis.tex
%*******************************************************************************
%****************************** Third Chapter **********************************
%*******************************************************************************
\chapter{Applications}

% **************************** Define Graphics Path **************************
\ifpdf
    \graphicspath{{Chapter3/Figs/Raster/}{Chapter3/Figs/PDF/}{Chapter3/Figs/}}
\else
    \graphicspath{{Chapter3/Figs/Vector/}{Chapter3/Figs/}}
\fi

In the first two chapters I have provided an overview of the theoretical basis of my work and placed it within the history of related approaches.
Here, I will demonstrate how the methods I develop can aid biological insight in a number of species domains.
To do so, I will use methods described in chapter 2 as implemented in {\tt vg}.

The small genome of \emph{Saccharomyces cerevisiae} and ready availability of sources for pangenomic data models made it very useful to my development of {\tt vg}.
I do so for a variety of pangenome constructions and also a variety of read lengths.

However, much interest in bioinformatics is with larger genomes, specifically human.
I use evaluations based on the human genome to validate the ability of {\tt vg} to scale to large genomes.
Through simulation and the analysis of real genomes, I show that {\tt vg} yields exactly the same quality of alignment as {\tt bwa mem} with runtime within five to ten-fold slower.
I develop a variation graph for the reference-guided genome assemblies from the HGSVC project and demonstrate the strong effect of reference bias in ChIP-seq.

One context where reference bias has very significant effects is in the analysis of ancient DNA (aDNA).
Here short reads and high intrinsic error rates encourage a high rate of reference bias.
I show that alignment against a pangenome graph ameliorates this issue.

{\tt vg} can be applied to any kind of variation graph.
To demonstrate the utility of this, I use de Bruijn assemblers to generate reference variation graphs.
I recreate a classical pangenomic analysis of core and accessory pangenome by analyzing the coverage of alignments mapped to an assembly graph built from 10 \emph{E. coli} strains.
{\tt vg} enables the full length alignment of reads to a complex assembly graph built from an arctic viral metagenome.

Finally, I demonstrate that the data models and indexes in {\tt vg} are capable of encoding splicing graphs, and that aligning to these splicing graphs allows the direct observation of the transcriptome.

\section{Yeast}
%*0.5p 0.5h*

\subsection{A SNP-based SGRP2 graph}
%*1p 1h*

\subsection{Cactus progressive assembly}
%*0.5p 0.5h*

\subsection{Constructing and comparing variation graphs from whole genome assemblies}
%*3p 2.5h*

\subsection{Long read mapping} % from paper / my experiments

\section{Human}
%*0.5p 0.5h*

\subsection{1000GP graph construction and indexing}
%*3p 3h*

\subsection{Simulations based on phased HG002}
\label{sec:1000GP_sim}
%*2p 2.5h*

\subsection{Whole genome variant calling experiments}
%*1p 2h*

\subsection{HGSVC from VCF and progressive alignment of human chromosomes}
%*2.5p 5h*

\subsection{CHiP-Seq}
%*1p 1h*


\section{Ancient DNA}
%*0.5p 0.5h*
With aDNA, read lengths are limited by the degredation of DNA due to environmental exposure.
Furthermore, degredation causes a high intrinsic error rate.
In combination these issues cause a strong bias against non-reference variation.
This has significant effect on population genetic inference and implications for many aDNA studies.

\subsection{Simulations with human origins panel}
%*1p 2h*

\subsection{Using 1000GP graph for samples from Martiniano et al 2016}
%*2p 4h*

\subsection{Evaluation of a high-coverage Botai sample}
%*2p 4h*


\section{Neoclassical bacterial pangenomics}
%*0.5p 0.5h*

\subsection{E. coli pangenome from illumina reads}
%*2p 4h*

\subsection{Evaluating the core and accessory pangenome}
%*2p 4h*


\section{Metagenomics}
%*0.5p 0.5h*

\subsection{Arctic viral metagenome}
%*2p 1.5h*

\subsection{Human gut microbiome}
%*2p 8h*


\section{RNAseq}
%*0.5p 0.5h*

\subsection{Yeast}
%*1p 2h*

\subsection{C. elegans}
%*2p 5h*

\subsection{Human}
%*2p 12h*


\section{Applications that I contributed to}

% ...
